### DISCUSSION TITLE
"low level" rendering textured triangles with support of stencil and render targets

### DISCUSSION THREAD

**Discussion by jeremyfa** [MonoGame/MonoGame#8184] [Graphics]
Hi 👋, I'm quite new to MonoGame (even if I have been following its developments for a while now, and the more recent creation of the Foundation).

I have a rather specific use case I'd like to explore which is running my open source 2D framework called **Ceramic** (https://ceramic-engine.com), written in the [Haxe](https://haxe.org) programming language, on top of **MonoGame** using Haxe's C# target. As silly as this idea sounds, I already successfully did something similar with Unity with pretty good results (the process, mostly using `Mesh` and `CommandBuffer` APIs to send vertex buffers to the GPU is [explained in this article](https://ceramic-engine.com/blog/ceramic-in-unity/)).

However, recent events around Unity (pricing debacle and loss of trust in general) made me reconsider this choice long term, so MonoGame became a very good candidate that supports many platforms (including consoles) and is more in line with my ideals of software (FOSS), where I could even contribute if needed.

So my question is: what is the best option in MonoGame to efficiently send textured triangles, or in other words vertex buffers, to the GPU, supporting render textures, stencil buffer and custom vertex attributes?

I explored the API docs a bit and I believe my best bet is `GraphicsDevice` and methods like `DrawIndexedPrimitives()` to send buffers, along with other methods of the same API to change the graphics device state like used textures, render target, shader....

Is that `GraphicsDevice` API the one I should look for? (asking for confirmation 😄), and are there edge cases I should be aware of? (like for example, stencil not working in some platforms, or multisampled render targets not working on some platforms, or anything I should pay attention to). I'll probably start writing a proof of concept in the following months!

Thanks a lot in advance for the hints!

**Comment by MetalMaxMX** [comment]
Hello! I am not a part of MonoGame development team, just to let you know, but I've read about the Ceramic engine and Haxe and it's certainly quite admirable of you to distance yourself from a proprietary game engine like Unity. I would like to know though but isn't Haxe removing support for C# backends? Which MonoGame is built on top of? I am referring to this commit in particular https://github.com/HaxeFoundation/haxe/pull/11551

Without Haxe's C# target, well, I think it's going to be difficult to use this.

**Reply by jeremyfa** [reply]
Hey @MetalMaxMX, yes current C# target of Haxe is going to be discontinued, but there is a [new C# target based on Reflaxe](https://github.com/SomeRanDev/reflaxe.CSharp) currently in the works, in which I'm personally involved. Meanwhile, there will still be the option to stick to Haxe 4 in order to use current C# target, no reason this would stop working, so C# from Haxe is not going away anytime soon. Sure there is work ahead to make the transition but I've seen worse 😄

**Reply by MetalMaxMX** [reply]
Hey there! So there were alternatives to that, yeah, makes sense now. Hope that the Reflaxe project you're involved in is successful for Haxe users! And yeah, one could also stick with Haxe 4 as you mentioned, it'll stay relevant for awhile too. 

I am afraid I cannot answer your MonoGame question as I am not really familiar with it, I was mostly curious about the Haxe's C# backend since that's a language I did happen to clone and build, and use a little bit. But still, hoping someone is able to answer your question!

Cheers!
MetalMaxMX

**Comment by ThomasFOG** [comment]
Hey Jérémy,

Yes, what you are looking for is the ```GraphicsDevice``` API. That's the API that is used by all the other parts of MonoGame to do its rendering. It's a pretty straightforward renderer that is built like an old school pipeline munching vertex/index buffers, and you can set shaders/textures/render targets on it. It's quite rigid but if you don't need to touch the swap chain or command buffers, that should do it.

Yes, some platforms have differences when it comes to multisampling, I don't have the full details in mind, but you can expect a couple of oddities.

Stencil should be working everywhere (on public platforms, at least).

And the rendering pipeline is limited to pixel and vertex shader stages.

**Reply by jeremyfa** [reply]
Sounds like the right thing indeed :)

Pixel and vertex shader only is fine, as well as the old school pipeline (which is pretty close to what I do with Ceramic anyway). And for the oddities, well, we'll see in time 😅.

Thanks for the hints! That should be more than enough to get started :)

**Comment by jeremyfa** [comment]
Just giving some news:

Things have moved a bit regarding exporting C# from Haxe: I am now actively working on a new C# target for Haxe 5 right on the compiler itself: https://github.com/HaxeFoundation/haxe/pull/12575, which already covers the language semantics and pass all tests.

There is no rush though, as Haxe 4 is still the stable path right now with C# support, but that means there is a clear path now beyond Haxe 4 for my use case.

On the other hand, I still didn't manage to dig into plugging MonoGame into Ceramic because I have been rather busy on other subjects, but still interested to do so!
